Fb380

joelaro04 opened the pull request:
Merged from master

Removed get_sync_data method

Updated filter_blocks documentation.

Run realtime listener in separate thread.

joelaro04 committed:
Replaced historical and realtime listeners with filter_blocks()

Merged from master

Removed get_sync_data method

Updated filter_blocks documentation.

Run realtime listener in separate thread.

kranar commented on augur/augur_client.py:
```diff
@@ -150,58 +158,41 @@ def event_signature_to_name_map(self):
+      filter (dict): Dictionary containing filter parameters.
+      handler (callback): Callback function for when an event log is found in
+                          the current block.
+      increment (int): Specifies the range of blocks from the current_block to
```
What does ```increment``` do?

joelaro04 replied:
`increment` is the value that determines the amount to increase the `toBlock` value in the filter dict by on each iteration. I'll change the name of this variable to `block_increment_size` to better describe what the variable does.

kranar replied:
The name is fine, I'm wondering what happens if your ```increment``` is instead ```end_block - start_block + 1```?

joelaro04 replied:
Filtering the blockchain with too large of a range resulted in the Web3py websocket timing out. I could try increasing the timeout/removing it and see how long that process is. My initial reasoning was that since it would timeout given the range of the entire blockchain, I could traverse it at a reasonable pace with a value like 1000 blocks/iteration.

kranar commented on augur/augur_client.py:
```diff
@@ -250,9 +245,10 @@ def load_market_id_from_order_id(self, order_id):
       self._is_open = True
     except (websockets.exceptions.InvalidURI,
-        websockets.exceptions.InvalidHandshake, OSError) as ws_error:
+            websockets.exceptions.InvalidHandshake, OSError) as ws_error:
```
Fix indentation.

kranar commented on augur/augur_client.py:
```diff
@@ -250,9 +245,10 @@ def load_market_id_from_order_id(self, order_id):
         self._contract_abi)
       self._event_signature_to_name_map = inverse_dict(
         self._event_name_to_signature_map)
+      self._latest_block = self._ethereum_client.eth.blockNumber
```
What is ```_latest_block``` used for?

joelaro04 replied:
`_latest_block` holds the latest Ethereum block at the time of the AugurClient startup. This was initially used to hold the end block for the historical listener, but does not do anything at the moment in the current branch as we supply the end block to `filter_blocks()`.

kranar commented on augur/augur_client.py:
```diff
@@ -322,33 +314,19 @@ def _contracts_from_abi(self, abi):
+    asyncio.set_event_loop(loop)
+
+    async def run_listener(handler, filter):
+      print('Watching for new blocks...')
```
Remove any debugging/printing when doing a pull request.

kranar commented:
Okay, your explanation of increment is good, thanks. Keep it.

On Tue, Feb 5, 2019 at 6:54 PM Joel Aro <notifications@github.com> wrote:

> *@joelaro04* commented on this pull request.
> ------------------------------
>
> In augur/augur_client.py
> <https://github.com/spiretrading/pyaugur/pull/5#discussion_r254094169>:
>
> > @@ -250,9 +245,10 @@ def load_market_id_from_order_id(self, order_id):
>          self._contract_abi)
>        self._event_signature_to_name_map = inverse_dict(
>          self._event_name_to_signature_map)
> +      self._latest_block = self._ethereum_client.eth.blockNumber
>
> _latest_block holds the latest Ethereum block at the time of the
> AugurClient startup. This was initially used to hold the end block for the
> historical listener, but does not do anything at the moment in the current
> branch as we supply the end block to filter_blocks().
>
> —
> You are receiving this because you commented.
> Reply to this email directly, view it on GitHub
> <https://github.com/spiretrading/pyaugur/pull/5#discussion_r254094169>,
> or mute the thread
> <https://github.com/notifications/unsubscribe-auth/APzFGbH8D3Nul1KP8a14S7wq_7WXF6Riks5vKhmZgaJpZM4akMgw>
> .
>


-- 
Kamal Mansouri
Computer Programmer
+1 647-648-0498

joelaro04 committed:
Removed debugging code, removed self._last_block

kranar commented on augur/augur_client.py:
```diff
@@ -150,58 +158,40 @@ def event_signature_to_name_map(self):
+    loop = asyncio.get_event_loop()
+    filter['fromBlock'] = hex(end_block)
+    filter.pop('toBlock', None)
+    ttt = threading.Thread(target=self._start_realtime_listener,
```
Try to do this without using another thread. Instead get the current thread's event loop and schedule to run ```_realtime_listener``` at some future point in time using ```asyncio.create_task```.

joelaro04 replied:
It looks like uvicorn destroys the asyncio event loop and starts a new one. We can tell uvicorn to use the same event loop that has the `filter_blocks` task on it by passing it to the `run` method:
```python
uvicorn.run(app, .., loop=asyncio.get_event_loop())
```

kranar replied:
Well man that would be awesome. See if it works and thanks for looking into that.

joelaro04 committed:
Removed threading. Added filter_blocks task.

Removed debugging code

joelaro04 committed:
Added sync_data public method

kranar reviewed (commented):
At what point do you get real time data?

joelaro04 commented:
> 
> 
> At what point do you get real time data?

`filter_blocks` will finish once the `end_block` is reached. The current implementation does not have realtime, but I can plug the code in for after `filter_blocks` reaches the `end_block`.

joelaro04 committed:
_filter_blocks will now watch for new blocks after parsing old events.

Removed debugging code.

kranar commented on augur/augur_client.py:
```diff
@@ -311,5 +311,16 @@ def _contracts_from_abi(self, abi):
+      filter['fromBlock'] = 'latest'
+      del filter['toBlock']
+      await self._send_request('eth_subscribe', websocket, ['logs', filter])
+      async for event in websocket:
```
Why is this for loop ```async``` but the one on line 315 is not ```async```?

joelaro04 replied:
The for loop on 315 is a list of events that the Web3py method `get_all_entries()` returns when it gets old events. The async on 323 waits on a response from the websocket. It uses `__aiter__` to iterate asynchronously through all messages that the `eth_subscribe` method outputs when a new event is found. Web3py has no method for 'subscribing' to Ethereum events, so I performed a JSON-RPC call to the node here.

kranar replied:
Makes sense.

kranar commented on augur/augur_client.py:
```diff
@@ -322,33 +303,24 @@ def _contracts_from_abi(self, abi):
+      self._addresses['Augur'])
+    for current_block in range(start_block, end_block, increment):
+      filter['fromBlock'] = current_block
+      filter['toBlock'] = current_block + increment
```
This line is a bug, consider ```start_block = 0```, ```end_block = 100```, ```increment = 1000```.

kranar commented on augur/augur_client.py:
```diff
@@ -322,33 +303,24 @@ def _contracts_from_abi(self, abi):
       for event in events:
-        on_event(event)
-      current_block += increment_block
+        await handler(event)
```
Does an ```await``` schedule the coroutine for immediate execution or does it schedule it for later?

joelaro04 replied:
It will schedule it for immediate execution. We already have `filter_blocks` scheduled for execution later on. Do we want to schedule handler to run at a later time as well? In the last commit, I removed async from the handler method, so handler calls will block the event loop, which is probably what we don't want.

kranar replied:
Handler calls should not block the event loop.

joelaro04 committed:
Handler is no longer expected to be a coroutine.
- Fixed 'toBlock' not being set if old events for loop is never run

Loads abi from json file

Testing with Ganache

Loaded Augur ABI from local file.

Re-added augur tester

Added abi_path to AugurClient parameters.

Handler will override Web3py filter's format_entry to better handle event entries

Removing non fb380 commits

Reset open()

Reset open(). Removed non-fb380 commits

joelaro04 committed:
Handler is now added to the event loop to prevent it from blocking

joelaro04 committed:
filter_blocks will now search only blocks between start_block and end_block. Future blocks will start a websocket to listen for

Added check for increment < 0.

Removed new link to abi.json

kranar reviewed (commented):
.

joelaro04 committed:
Moved code to validate _filter_blocks parameters to filter_blocks.

Moved event_block to inside new_blocks

kranar reviewed (changes requested):
.

joelaro04 committed:
Removed logs subscription. events_filter will output all events that match in that block

Changed exception message

kranar commented on augur/augur_client.py:
```diff
@@ -181,11 +181,9 @@ def filter_blocks(self, start_block, end_block, filter, handler,
     self._require_is_open()
     if (start_block < 0) or (end_block < 0) or (increment < 0):
-      raise IOError('Block numbers and increments should be greater than 0.')
+      raise IOError('Block numbers or increment cannot be less than 0.')
```
The error messages still don't match the check being performed. Ensure that they are consistent.

joelaro04 replied:
Is the issue here that the error message does not identify which parameter is raising the exception? Instead of having one check for all 3 parameters I changed it to each parameter raising their own exception.

kranar replied:
It's a matter of being very clear and precise in your messages. You say ```start_block``` must be less than ```end_block```, but you only check if ```start_block``` is greater than ```end_block```. Pick a set of semantics and ensure it's applied consistently throughout the entire function.

kranar commented on augur/augur_client.py:
```diff
@@ -181,11 +181,9 @@ def filter_blocks(self, start_block, end_block, filter, handler,
     '''
     self._require_is_open()
     if (start_block < 0) or (end_block < 0) or (increment < 0):
```
Can ```increment``` be 0?

joelaro04 replied:
`increment` is used to increase the blocks to search in. Setting it to 0 would not allow `_filter_blocks` to crawl through the blockchain. However, an `increment` of 0 could be used for a small range of blocks between `fromBlock` and `toBlock`, and would not need to enter the for loop.

kranar replied:
What is the formal definition of a small range for which an increment of 0 works?

joelaro04 replied:
There is no formal definition for what a small range for the increment to be, since the user can define how fast they want to search through the blockchain. Before finding out that there was a payload size limit on the websocket, I thought the timeout was a result of the range between `fromBlock` and `toBlock` was too big and processing the data was taking too long. To workaround the timeout, I set a base `increment` value of 1000 so that the websocket would not timeout and it would search the blockchain at a reasonable pace (rather than incrementing by 1 block at a time).
Since the timeout is a result of the payload size being larger than the default websocket size limit, `_filter_blocks` should be able to receive an `increment` of 0. A small range then, would be one that would return a payload from the websocket that is less than the set websocket's payload size.

kranar replied:
Carefully trace through your code to determine what it means for an ```increment``` to be ```0``` and fix the bug.

kranar commented on augur/augur_client.py:
```diff
@@ -181,11 +181,9 @@ def filter_blocks(self, start_block, end_block, filter, handler,
       raise IOError('start_block must be less than the end_block.')
-    if start_block > self._ethereum_client.eth.blockNumber:
-      raise IOError('start_block has not been mined yet.')
     loop = asyncio.get_event_loop()
```
Eliminate the use of the variable, a variable that is only ever used once is not worth using at all.

kranar commented on augur/augur_client.py:
```diff
@@ -316,39 +314,42 @@ def _contracts_from_abi(self, abi):
     # execution. It will then listen for new blocks until the desired end_block
     # has been reached.
-    filter['address'] = self._ethereum_client.toChecksumAddress(
+    filter_params = filter
```
Make a copy of the ```filter``` before you mutate it.

joelaro04 committed:
Changed start_block > end_block IOError message.  Added check for increment == 0.

Updated increment check to include both less than and equal to

Reset abi_url back to old url

kranar commented on augur/augur_client.py:
```diff
@@ -180,13 +180,16 @@ def filter_blocks(self, start_block, end_block, filter, handler,
-      raise IOError('Block numbers or increment cannot be less than 0.')
+    if start_block < 0:
+      raise IOError('start_block cannot be less than 0.')
+    if end_block < 0:
```
This check is redundant, it is a strictly weaker check than the one performed on line 189.

kranar commented on augur/augur_client.py:
```diff
@@ -314,7 +317,7 @@ def _contracts_from_abi(self, abi):
     # execution. It will then listen for new blocks until the desired end_block
     # has been reached.
-    filter_params = filter
+    filter_params = dict(filter)
```
Making the copy here is too late, the caller might have already modified the filter before execution reaches line 320.

joelaro04 replied:
I have updated this so that the filter is copied before an asyncio task is created. This way the filter is copied after the parameters have been validated and before running `_filter_blocks`.

joelaro04 committed:
Removed redundant check. Moved copied filter to filter_blocks().

Removed unused copy_filter parameter
